### PR TITLE
debezium/dbz#2088 Add SSH config parser for host discovery

### PR DESCRIPTION
**Pull request by d1vyanshu-kumar**
Fixes debezium/dbz#2088

## Description

Adds `SshConfigParser`, a stateless `@ApplicationScoped` bean that reads anSSH config file and returns a list of `SshHostEntry` records with the connection details (alias, hostname, user, port, identity file).

The parser handles the standard `ssh_config(5)` format: both whitespace and `=` delimiters, inline comments, quoted paths with spaces, case-insensitive keywords, and wildcard entries (`Host *`, `Host db-*`) which are skipped
since they don't represent actual hosts.

No existing files were modified. No new dependencies were added - parsing uses only `java.nio.file` and standard `String` methods.

The `SshConfigWatcherService` ([next sub-issue](https://github.com/debezium/dbz/issues/2090)) will `@Inject` this parser and call `parse(Path)` whenever the SSH config file changes on disk.

24 unit tests cover all parsing rules, including edge cases (malformed lines, out-of-range ports, duplicate aliases, alias-only entries).

## PR Checklist

- [X] I have read the [contribution guidelines](https://github.com/debezium/debezium/blob/main/CONTRIBUTING.md) and the [governance document](https://github.com/debezium/governance/blob/main/GOVERNANCE.md) on PR expectations.
- [X] Minimal changes to code not directly related to your change (e.g. no unnecessary formatting changes or refactoring to existing code)
- [X] One feature/change per PR unless tightly coupled
- [X] Do a rebase on upstream `main`

### CODE REVIEW

**Comment by d1vyanshu-kumar** [comment]
@kmos @mfvitale, could you please take a look at this PR and let me know if any changes are needed?

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    /**
+     * Parses the SSH config file at the given path.
+     *
+     * @param configFilePath path to the SSH config file
+     * @return list of parsed host entries, never null
+     * @throws IOException if the file cannot be read
+     */
+    public List<SshHostEntry> parse(Path configFilePath) throws IOException {
+        logger.debugv("Parsing SSH config file: {0}", configFilePath);
+        String content = Files.readString(configFilePath);
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Maybe you can use the `Files.readAllLines(configFilePath)`.

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+        String currentAlias = null;
+        String currentHostname = null;
+        String currentUser = null;
+        int currentPort = DEFAULT_PORT;
+        String currentIdentityFile = null;
+
+        try (BufferedReader reader = new BufferedReader(new StringReader(content))) {
+            String line;
+            while ((line = reader.readLine()) != null) {
+                String trimmed = line.trim();
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: `strip()` should be more powerful in this context.

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+
+                // Rule 1: skip blank lines and full-line comments
+                if (trimmed.isEmpty() || trimmed.charAt(0) == '#') {
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: maybe you can wrap the `trimmed.isEmpty() || trimmed.charAt(0) == '#'` in a method with self explaining name and remove the comment above. WDYT?

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+
+        return result;
+    }
+
+    /**
+     * Finds the index of the first delimiter (whitespace or '=') in the line.
+     * Returns -1 if no delimiter is found.
+     */
+    private int findDelimiterIndex(String line) {
+        for (int i = 0; i < line.length(); i++) {
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Can you use the `IntStream.range` here?

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+                keyword = trimmed.substring(0, splitIndex).trim().toLowerCase();
+                // Skip past the delimiter character and trim
+                String rawValue = trimmed.substring(splitIndex + 1).trim();
+
+                // Handle mixed delimiter: "Key = Value" splits at the space, leaving "= Value".
+                // Strip the leading '=' and re-trim so all three forms produce the same value:
+                // "Port 22" → rawValue = "22"
+                // "Port=22" → rawValue = "22"
+                // "Port = 22" → rawValue = "= 22" → strip '=' → "22"
+                if (!rawValue.isEmpty() && rawValue.charAt(0) == '=') {
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Could you add a constant for `=` and also for `#` above?

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+                    rawValue = rawValue.substring(1).trim();
+                }
+
+                if (rawValue.isEmpty()) {
+                    logger.warnv("Malformed line in SSH config (empty value): {0}", trimmed);
+                    continue;
+                }
+
+                // Rule 3: strip inline comments (unquoted '#')
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: I think you can remove the comment since the method name is self exaplaining

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+                value = stripInlineComment(rawValue);
+
+                // Rule 4: strip surrounding quotes
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Same as above.

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+                value = stripQuotes(value);
+
+                // Rule 5: keyword is already lowercased above
+
+                if ("host".equals(keyword)) {
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Could `"host".equals(keyword)` be in a method called `isHeadingLine` or something similar?

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+        logger.debugv("Found {0} host entries in SSH config", entries.size());
+        return entries;
+    }
+
+    /**
+     * Parses the given SSH config content string into host entries.
+     * Package-visible for unit test access.
+     */
+    List<SshHostEntry> parseContent(String content) {
+        List<SshHostEntry> result = new ArrayList<>();
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Maybe you can use a more explicative name instead of `result`. Something like `hostEntries`?

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+        return value;
+    }
+
+    /**
+     * Returns true if the host alias contains wildcard characters (*, ?, !).
+     */
+    private boolean isWildcard(String alias) {
+        for (int i = 0; i < alias.length(); i++) {
+            char c = alias.charAt(i);
+            if (c == '*' || c == '?' || c == '!') {
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Could you add costants?

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+            return value.substring(1, value.length() - 1);
+        }
+        return value;
+    }
+
+    /**
+     * Returns true if the host alias contains wildcard characters (*, ?, !).
+     */
+    private boolean isWildcard(String alias) {
+        for (int i = 0; i < alias.length(); i++) {
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Also here maybe you can try to uses java streams?

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+                        currentHostname = null;
+                        currentUser = null;
+                        currentPort = DEFAULT_PORT;
+                        currentIdentityFile = null;
+                        continue;
+                    }
+
+                    // Rule 11: duplicate alias detection
+                    if (!seenAliases.add(value)) {
+                        logger.warnv("Duplicate Host alias found: {0}", value);
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Should we skip it?

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Parses a port string into an integer, falling back to the default port (22)
+     * if the value is not a valid integer or is outside the allowed range.
+     */
+    private int parsePort(String value) {
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java:251]: In case we are not able to parse the port or if the port is outside the valid range i think we should skip the entry instead using the default port. WDYT?

File: debezium-platform-conductor/src/test/java/io/debezium/platform/environment/host/discovery/SshConfigParserTest.java [hunk 14]
```diff
@@ -0,0 +1,477 @@
...
+class SshConfigParserTest {
+
+    private SshConfigParser parser;
+
+    @BeforeEach
+    void setUp() {
+        parser = new SshConfigParser(Logger.getLogger(SshConfigParser.class));
+    }
+
+    // ── TEST 1: Standard whitespace delimiter ──────────────────────────────
```
Comment (mfvitale) [inline debezium-platform-conductor/src/test/java/io/debezium/platform/environment/host/discovery/SshConfigParserTest.java]: You can remove this comments.

File: debezium-platform-conductor/src/test/java/io/debezium/platform/environment/host/discovery/SshConfigParserTest.java [hunk 15]
```diff
@@ -0,0 +1,477 @@
...
+
+    @Test
+    void testStandardWhitespaceDelimiterEntry() {
+        String config = """
```
Comment (mfvitale) [inline debezium-platform-conductor/src/test/java/io/debezium/platform/environment/host/discovery/SshConfigParserTest.java]: WDYT to add this directly in files under `resourses` folder so you can use directly the `parse` method?

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+        List<SshHostEntry> hostEntries = new ArrayList<>();
+        Set<String> seenAliases = new HashSet<>();
+
+        String currentAlias = null;
+        String currentHostname = null;
+        String currentUser = null;
+        int currentPort = DEFAULT_PORT;
+        String currentIdentityFile = null;
+
+        for (String line : lines) {
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Maybe you can use a bit of Java streams here. 

You could encapsulate the parsing of a single line a a method and the use the `map` and `toList` to get the list of parsed lines.

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+     * '#' that is not inside double quotes.
+     *
+     * <p>Example: {@code "deploy # production"} becomes {@code "deploy"}.
+     * <p>Example: {@code "\"path with # in it\""} preserves the '#' inside quotes.
+     */
+    private String stripInlineComment(String value) {
+        boolean inQuotes = false;
+        for (int i = 0; i < value.length(); i++) {
+            char c = value.charAt(i);
+            if (c == '"') {
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Another constant?

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java (hunk 16 above)
Comment (d1vyanshu-kumar) [reply debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Done! Refactored parseLines into two phases: a parseLine(String) method that extracts keyword-value pairs wrapped in Optional, streamed via map + flatMap(Optional::stream) + toList, and a separate buildHostEntries method for the stateful block grouping.

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java (hunk 17 above)
Comment (d1vyanshu-kumar) [reply debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Good catch, added` QUOTE_CHAR`, missed it when adding the other delimiter constants.

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+
+                currentAlias = parsed.value();
+                currentHostname = null;
+                currentUser = null;
+                currentPort = DEFAULT_PORT;
+                currentIdentityFile = null;
+            }
+            else if (currentAlias != null) {
+                switch (parsed.keyword()) {
+                    case "hostname":
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Maybe you can use the arrow syntax here.

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java [hunk 19]
```diff
@@ -0,0 +1,300 @@
...
+     */
+    private List<SshHostEntry> buildHostEntries(List<ParsedLine> parsedLines) {
+        List<SshHostEntry> hostEntries = new ArrayList<>();
+        Set<String> seenAliases = new HashSet<>();
+
+        String currentAlias = null;
+        String currentHostname = null;
+        String currentUser = null;
+        int currentPort = DEFAULT_PORT;
+        String currentIdentityFile = null;
```
Comment (mfvitale) [inline debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Maybe there is still room for improvement here. 

You could add these to a inner class, adding to them the "reset" logic and also the `toSshHostEntry` conversion.

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java (hunk 18 above)
Comment (d1vyanshu-kumar) [reply debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: Done! switched to arrow-case syntax.

File: debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java (hunk 19 above)
Comment (d1vyanshu-kumar) [reply debezium-platform-conductor/src/main/java/io/debezium/platform/environment/host/discovery/SshConfigParser.java]: @mfvitale, I think that's a good call! I extracted a `HostBlock` inner class with `reset()`, `startNew()`, and `toSshHostEntry()`. Eliminates the repeated field assignments. Please take a look and let me know if it looks good or if there is still room for improvement!

**Comment by d1vyanshu-kumar** [comment]
@mfvitale it seems like the current CI failure is due to an upstream compilation issue in `debezium-connector-db2 `on the main branch, which is unrelated to these changes. Could you please take a look when you have a moment?

**Review by mfvitale — APPROVED** [review]

**Comment by mfvitale** [comment]
@d1vyanshu-kumar Applied, thanks!
